aws-common/src/main/java/org/apache/druid/common/aws/LazyFileSessionCredentialsProvider.java

```diff
@@ -32,14 +32,10 @@ public LazyFileSessionCredentialsProvider(AWSCredentialsConfig config)
     this.config = config;
   }
 
-  private FileSessionCredentialsProvider getUnderlyingProvider()
+  private synchronized FileSessionCredentialsProvider getUnderlyingProvider()
```

Comment:
The former code here is indeed buggy, however putting `synchronized` on `getUnderlyingProvider()` is not how this method should be fixed.

Please make the following:
 1. Declare `config` field `final`
 2. Add a comment noting that `FileSessionCredentialsProvider` is safely initialized because it contains a `final` field `sessionCredentials`. Also a comment should be added on that field noting that `final` modifier should not be removed from it, because it has concurrency implications in `LazyFileSessionCredentialsProvider`.
 3. In `getUnderlyingProvider()`, `provider` should be read into a local variable, local variable updated in the `synchronized` block along with the field, and the local is returned in the last line of the method. This is because currently the second read of `provider` in the last line of the method could return `null` even if the first read in the first line returned non-null and the `synchronized` block was not entered.

Reply:
@leventov do you mean that?

`  private final AWSCredentialsConfig config;
  @Nullable
  private FileSessionCredentialsProvider provider;

  public LazyFileSessionCredentialsProvider(AWSCredentialsConfig config)
  {
    this.config = config;
  }

  //FileSessionCredentialsProvider is safely initialized because it contains a final field sessionCredentials.
  private synchronized FileSessionCredentialsProvider getUnderlyingProvider()
  {
    FileSessionCredentialsProvider syncedProvider = provider;
    if (syncedProvider == null) {
      synchronized (config) {
        syncedProvider = new FileSessionCredentialsProvider(config.getFileSessionCredentials());
         provider = syncedProvider;
      }
    }
    return syncedProvider;
  }`

Reply:
1) No `synchronized` on the method
2) There should be another check inside the `synchronized` block (because currently it's not **double** checked locking :)
3) I would put the comment before the `provider = syncedProvider;` line:
```
// The provider field doesn't need to volatile because FileSessionCredentialsProvider
// is safely initialized because it contains a final field sessionCredentials.
```
Added "The provider field doesn't need to volatile" because that's what specifically needs to be explained here.

Reply:
Also for the comment to be accurate, this PR should be merged not earlier than this one: #6664

Reply:
This post helps make sense of what @leventov is saying: https://shipilev.net/blog/2014/safe-public-construction/. I admit I had to read it before understanding what was meant by the comment "FileSessionCredentialsProvider is safely initialized because it contains a final field sessionCredentials".

@leventov - am I understanding you right - are you suggesting that we should rely on the fact that an object with at least one final field will always be safely initialized, and skip the volatile? i.e. in the terms used in the post linked above, you are suggesting it's best to do "Unsafe Local DCL + Safe Singleton" in this case?

If so, I don't feel super comfortable about that. The reason is that it is easy to screw it up: removing a final field in a different class (the class being initialized) breaks the holder class's correctness. Comments help with this problem, but IMO, it makes life easier to make the `provider` field volatile, which makes the correctness obvious from just looking at the container class (in this case, LazyFileSessionCredentialsProvider). Limiting the scope of what a maintainer needs to consider to a single file is a good thing, since it makes future maintenance easier and less error-prone.

Separately from that, it is not clear to me that this behavior is in fact guaranteed. Please educate me if I am missing something (I haven't studied the relevant sections of the Java spec) but it seems like the "do a totally safe initialization if _any_ final is written" behavior is just an implementation detail of a particular VM and not something guaranteed. If that is right, then to get guaranteed safe initialization of an entire class, _all_ fields should be final (or volatile).

Long story short, in this (non-perf-critical) case I am advocating for "Safe DCL", meaning the code below. And in performance critical code, "Unsafe Local DCL + Safe Singleton" could make sense, although it's worth verifying. The linked post above is 4 years old and it just did one set of tests, but, in that test that pattern didn't run any faster than "Safe DCL" or "Safe Local DCL" on x86.

```java
private final AWSCredentialsConfig config;

@Nullable
private volatile FileSessionCredentialsProvider provider;

public LazyFileSessionCredentialsProvider(AWSCredentialsConfig config)
{
  this.config = config;
}

private synchronized FileSessionCredentialsProvider getUnderlyingProvider()
{
  if (provider == null) {
    synchronized (config) {
      if (provider == null) {
        provider = new FileSessionCredentialsProvider(config.getFileSessionCredentials());
      }
    }
  }
  return provider;
}
```

Reply:
@gianm synchronized on a final field within a synchronized method seems doesn't necessary?

Reply:
Sorry, & thanks, that was a copy-and-paste error. I didn't mean to have the synchronized keyword on the method. I edited the comment to fix it. Synchronizing a method is just like wrapping the body in a `synchronized (this) { ... }` block, which wouldn't be useful here.

Reply:
@gianm general suggestion looks good.

1) We could still skip volatile when the initialized object is obviously immutable, such as a String and and immutable collection.
2) Even with volatile, introducing local variable is more accurate performance-wise and should be done.

> If that is right, then to get guaranteed safe initialization of an entire class, all fields should be final (or volatile).

Correction - volatile on the initialized object's fields does *not* guarantee you anything here. In some sense, `volatile` is *less strong* than `final`. See https://shipilev.net/blog/2016/close-encounters-of-jmm-kind/#wishful-volatiles-are-finals

Reply:
@leventov 

When you say:

> 2. There should be another check inside the synchronized block (because currently it's not double checked locking :)

if you mean that:

```java  
 //FileSessionCredentialsProvider is safely initialized because it contains a final field sessionCredentials.
  private FileSessionCredentialsProvider getUnderlyingProvider()
  {
    FileSessionCredentialsProvider syncedProvider = provider;
    if (syncedProvider == null) {
      synchronized (config) {
        if (syncedProvider == null) {
          syncedProvider = new FileSessionCredentialsProvider(config.getFileSessionCredentials());
          // The provider field doesn't need to be volatile because FileSessionCredentialsProvider
          // is safely initialized because it contains a final field sessionCredentials.
          provider = syncedProvider;
        }
      }
    }
    return syncedProvider;
  }
```

This leads to unnecessary check for `syncedProvider` since its value is always null. Below seems to be enough?

```java
  //FileSessionCredentialsProvider is safely initialized because it contains a final field sessionCredentials.
  private FileSessionCredentialsProvider getUnderlyingProvider() {
    FileSessionCredentialsProvider syncedProvider = provider;
    if (syncedProvider == null) {
      synchronized (config) {
        syncedProvider = new FileSessionCredentialsProvider(config.getFileSessionCredentials());
        // The provider field doesn't need to be volatile because FileSessionCredentialsProvider
        // is safely initialized because it contains a final field sessionCredentials.
        provider = syncedProvider;

      }
    }
    return syncedProvider;
  }
```

Reply:
You need to re-read the field into the local variable inside the synchronized block. That's the whole point of double-checked locking.

Reply:
Also since `volatile` you don't need that "safely initialized" comment on the field write line, but you need to write a similar comment on the field declaration, saying that the field is declared volatile in order to ensure safe publication of the object without worrying about final modifiers on the fields of the created object.

Reply:
Here is the implementation with reference to _Effective Java, Second Edition Item 71: Use lazy initialization judiciously_:

```java
  private final AWSCredentialsConfig config;

  /* The field is declared volatile in order to ensure safe publication of the object
   without worrying about final modifiers on the fields of the created object */
  @Nullable
  private volatile FileSessionCredentialsProvider provider;

  public LazyFileSessionCredentialsProvider(AWSCredentialsConfig config)
  {
    this.config = config;
  }

  //FileSessionCredentialsProvider is safely initialized because it contains a final field sessionCredentials.
  private FileSessionCredentialsProvider getUnderlyingProvider() {
    FileSessionCredentialsProvider syncedProvider = provider;
    if (syncedProvider == null) {
      synchronized (config) {
        syncedProvider = provider;
        if (syncedProvider == null) {
          syncedProvider = new FileSessionCredentialsProvider(config.getFileSessionCredentials());
          provider = syncedProvider;
        }
      }
    }
    return syncedProvider;
  }
```

Reply:
👍 except the final thing, the comment should be a javadoc comment:
```java
/**
 * The field is declared volatile in order to ensure safe publication of the object
 * in {@link #getUnderlyingProvider()} without worrying about final modifiers
 * on the fields of the created object
 */
```

Reply:
This line should also be removed:
```java
//FileSessionCredentialsProvider is safely initialized because it contains a final field sessionCredentials.
```

Reply:
To extend Gian's point from https://github.com/apache/incubator-druid/pull/6662#discussion_r237013157, there is another reason why I think that double-checked locking without volatile should be prohibited in Druid entirely. When some code is added that bypasses the initialization path, and only checks that the field is initialized, a-la
```java
if (lazilyInitializedField != null) {
  doSomethingWith(lazilyInitializedField);
}
```
There is a race and possibility of NPE. The second read of the field might result in null, despite the first read returned non-null and the check was passed. The correct code is:
```java
Foo lazilyInitializedField = this.lazilyInitializedField;
if (lazilyInitializedField != null) {
  doSomethingWith(lazilyInitializedField);
}
```
But I think it's unrealistic to expect this level of attention to details and discipline from all developers and reviewers.